test/java/org/chromium/net/CronetEngineBuilderTest.java

```diff
@@ -0,0 +1,155 @@
+  }
+
+  /**
+   * Tests the correct ordering of the providers. The platform provider should be
+   * the last in the list. Other providers should be ordered by placing providers
+   * with the higher version first.
+   */
+  @Test
+  @SmallTest
+  @Feature({"Cronet"})
+  public void testProviderOrdering() {
+    final CronetProvider[] availableProviders = new CronetProvider[] {
+        new FakeProvider(getContext(), PROVIDER_NAME_APP_PACKAGED, "99.77", true),
+        new FakeProvider(getContext(), PROVIDER_NAME_FALLBACK, "99.99", true),
+        new FakeProvider(getContext(), "Some other provider", "99.88", true),
+    };
+
+    ArrayList<CronetProvider> providers = new ArrayList<>(Arrays.asList(availableProviders));
+    List<CronetProvider> orderedProviders =
+        CronetEngine.Builder.getEnabledCronetProviders(getContext(), providers);
+
+    // Check the result
+    assertEquals(availableProviders[2], orderedProviders.get(0));
+    assertEquals(availableProviders[0], orderedProviders.get(1));
+    assertEquals(availableProviders[1], orderedProviders.get(2));
+  }
+
+  /**
+   * Tests that the providers that are disabled are not included in the list of available
+   * providers when the provider is selected by the default selection logic.
+   */
+  @Test
+  @SmallTest
+  @Feature({"Cronet"})
+  public void testThatDisabledProvidersAreExcluded() {
+    final CronetProvider[] availableProviders = new CronetProvider[] {
+        new FakeProvider(getContext(), PROVIDER_NAME_FALLBACK, "99.99", true),
+        new FakeProvider(getContext(), PROVIDER_NAME_APP_PACKAGED, "99.77", true),
+        new FakeProvider(getContext(), "Some other provider", "99.88", false),
+    };
+
+    ArrayList<CronetProvider> providers = new ArrayList<>(Arrays.asList(availableProviders));
+    List<CronetProvider> orderedProviders =
+        CronetEngine.Builder.getEnabledCronetProviders(getContext(), providers);
+
+    assertEquals("Unexpected number of providers in the list", 2, orderedProviders.size());
+    assertEquals(PROVIDER_NAME_APP_PACKAGED, orderedProviders.get(0).getName());
+    assertEquals(PROVIDER_NAME_FALLBACK, orderedProviders.get(1).getName());
+  }
+
+  private void assertVersionIsHigher(String s1, String s2) {
+    assertEquals(1, CronetEngine.Builder.compareVersions(s1, s2));
+  }
+
+  private void assertVersionIsLower(String s1, String s2) {
+    assertEquals(-1, CronetEngine.Builder.compareVersions(s1, s2));
+  }
+
+  private void assertVersionIsEqual(String s1, String s2) {
+    assertEquals(0, CronetEngine.Builder.compareVersions(s1, s2));
+  }
+
+  private void assertIllegalArgumentException(String s1, String s2) {
+    try {
+      CronetEngine.Builder.compareVersions(s1, s2);
+    } catch (IllegalArgumentException e) {
+      // Do nothing. It is expected.
+      return;
+    }
+    fail("Expected IllegalArgumentException");
+  }
+
+  // TODO(kapishnikov): Replace with a mock when mockito is supported.
```

Comment:
Mockito is supported in Envoy Mobile tests - probably a good time to make that todo go away.